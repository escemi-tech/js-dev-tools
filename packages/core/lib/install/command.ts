import { join, resolve } from "path";

import { readPackageJson } from "../services/packageJson";
import { executeMigrations } from "./migration";

export function install({ cwd, dir = "." }: { cwd: string; dir: string }): void {
  const packageName = readPackageJson(join(__dirname, "../..")).name;

  // Ensure that we're not trying to install outside cwd
  const absoluteProjectDir = resolve(cwd, dir);
  if (!absoluteProjectDir.startsWith(cwd)) {
    throw new Error(".. not allowed");
  }

  // Run installation migration
  const packageJson = readPackageJson(absoluteProjectDir);
  const currentVersion = packageJson.tsDevTools?.version;

  if (currentVersion) {
    console.info(`Updating ${packageName} installation...`);
  } else {
    console.info(`Installing ${packageName}...`);
  }

  executeMigrations(absoluteProjectDir, currentVersion);

  console.info(`Installation done!`);
}
