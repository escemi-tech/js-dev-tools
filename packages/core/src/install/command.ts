import { resolve } from "path";

import { DuplicateDependenciesService } from "../services/DuplicateDependenciesService";
import { MigrationsService } from "../services/MigrationsService";
import { PackageJson } from "../services/PackageJson";

export async function install({ cwd, dir = "." }: { cwd: string; dir: string }): Promise<void> {
  const tsDevToolsRootPath = resolve(__dirname, "../..");

  const packageName = PackageJson.fromDirPath(tsDevToolsRootPath).getPackageName();

  // Ensure that we're not trying to install outside cwd
  const absoluteProjectDir = resolve(cwd, dir);
  if (!absoluteProjectDir.startsWith(cwd)) {
    throw new Error(".. not allowed");
  }

  // Run installation migration
  const currentVersion = PackageJson.fromDirPath(absoluteProjectDir).getTsDevToolsVersion();

  if (currentVersion) {
    console.info(`Updating ${packageName} installation...`);
  } else {
    console.info(`Installing ${packageName}...`);
  }

  await MigrationsService.executeMigrations(tsDevToolsRootPath, absoluteProjectDir, currentVersion);

  DuplicateDependenciesService.duplicateDependencies(tsDevToolsRootPath, absoluteProjectDir);

  console.info(`Installation done!`);
}
