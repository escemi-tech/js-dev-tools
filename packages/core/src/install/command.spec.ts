import { createTestProjectDir, removeTestProjectDir } from "../tests/utils";
import { install } from "./command";

describe("Install command", () => {
  let testProjectDir: string;

  beforeAll(() => {
    testProjectDir = createTestProjectDir(__filename);
  });

  afterAll(() => {
    removeTestProjectDir(__filename);
  });

  it("should run installation without error", async () => {
    const result = await install({ cwd: testProjectDir, dir: "." });
    expect(result).toBeUndefined();
  });
});
