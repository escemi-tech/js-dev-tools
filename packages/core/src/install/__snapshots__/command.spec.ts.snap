// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Install command should run fresh installation without error 1`] = `
"Installing @ts-dev-tools/core...
Checking for duplicate dev dependencies...
Check for duplicate dev dependencies done!
Installation done!"
`;

exports[`Install command should run update without error 1`] = `
"Updating @ts-dev-tools/core installation...
Checking for duplicate dev dependencies...
Check for duplicate dev dependencies done!
Installation done!"
`;
